docs/proposals/logdir-volume.md

```diff
@@ -0,0 +1,190 @@
+
+```
+* EmptyDir: <kubelet_root_dir>/pods/<pod_id>/volumes/kubernetes.io~empty-dir/<volume_name>
+* LogDir:   <kubelet_root_dir>/pods/<pod_id>/volumes/kubernetes.io~log-dir/<volume_name>/<pod_full_name>
+```
+
+As we can see, `emptyDir` volume add a directory level: `pod_full_name`, the purpose is to make logging agent(e.g.Fluentd) easy to add tag by `pod_full_name`.
+
+Add the following content to `api.VolumeSource` struct:
+
+```
+type VolumeSource struct {	
+    ...
+	// LogDir represents a log directory that preserves a pod's log files.
+	LogDir *LogDirVolumeSource `json:"logDir,omitempty"`
+	...
+}
+
+// LogDirVolumeSource represents an log directory for a pod.
+type LogDirVolumeSource struct {
+	// TODO: Support log rotation and maximum log file sizes, etc.
+	// Optional: what type of storage medium should back this directory.
+	// The default is "" which means to use the node's default medium.
+	Medium StorageMedium `json:"medium"`
+}
+```
+
+Add  `logDirPlugin` to implementation `volume.VolumePlugin` interface:
+
+```
+type logDirPlugin struct {
+	host volume.VolumeHost
+}
+
+const (
+	logDirPluginName = "kubernetes.io/log-dir"
+)
+
+func (plugin *logDirPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *logDirPlugin) Name() string {
+	return logDirPluginName
+}
+
+func (plugin *logDirPlugin) CanSupport(spec *volume.Spec) bool {
+	if spec.VolumeSource.LogDir != nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *logDirPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, pod, mounter, &realMountDetector{mounter}, opts, newChconRunner())
+}
+
+func (plugin *logDirPlugin) NewCleaner(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newCleanerInternal(volName, podUID, mounter, &realMountDetector{mounter})
+}
+```
+
+## Integrated with Fluentd
+
+We can use Fluentd to collect log files in `LogDir` volume. Fluentd should be installed on each Kubernetes node, and it will watch `LogDir` volume's path, the Fluentd's configuration as follows(we use kubelet_root_dir's defalut value /var/lib/kubelet, ):
+
+```
+<source>
+  type tail
+  format none
+  time_key time
+  path /var/lib/kubelet/pods/*/volumes/kubernetes.io~log-dir/**/*.log
+  pos_file /lib/pods.log.pos
+  time_format %Y-%m-%dT%H:%M:%S
+  tag reform.*
+  read_from_head true
+</source>
+
+<match reform.**>
+  type record_reformer
+  enable_ruby true
+  tag kubernetes.${hostname}.${tag_suffix[10]}
+</match>
+
+<match **>
+  type stdout
+</match>
+```
+
+The Fluentd will tail any files in the `LogDir` volume, and add tag `kubernetes.<node_host_name>.<pod_full_name>.<file_name>`. We only print the logs to stdout,  also can forward to logging storage endpoint(e.g Elasticsearch).
```

Comment:
It's not the end of the world, but for what it's worth our current Fluentd configurations also include the container name.

There's also the interesting future option of being able to use this to modify legacy file names by being able to map a container filename to a different filename outside the container, but that's way less important than the feature as a whole.